core/arch/arm/kernel/ftrace.c

```diff
@@ -0,0 +1,150 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (C) 2019, Linaro Limited
+ */
+
+#include <assert.h>
+#include <kernel/ftrace.h>
+#include <kernel/panic.h>
+#include <kernel/thread.h>
+#include <kernel/tee_ta_manager.h>
+#include <mm/mobj.h>
+#include <optee_rpc_cmd.h>
+#include <printk.h>
+#include <string.h>
+#include <trace.h>
+#include "elf_load_dyn.h"
+
+#define MIN_FTRACE_BUF_SIZE	1024
+#define MAX_HEADER_STRLEN	128
+
+void ta_fbuf_init(struct ta_head *ta_head, struct tee_ta_session *s,
+		  struct elf_load_state *state)
+{
+	struct __ftrace_info *finfo = NULL;
+	struct ftrace_buf *fbuf = NULL;
+	size_t fbuf_size = 0;
+	uintptr_t val = 0;
+	int count = 0;
+	TEE_Result res = TEE_SUCCESS;
+
+	res = elf_resolve_symbol(state, "__ftrace_info", &val);
+	if (res != TEE_SUCCESS || !val) {
+		DMSG("ftrace info not found");
+		return;
+	}
+
+	finfo = (struct __ftrace_info *)(val + (uintptr_t)ta_head);
```

Comment:
What does `ta_head` have to do here?

How did you test this PR? Have you run the whole `xtest` suite with ftrace enabled?
I'm asking because I am having all sorts of problems, such as: TA panic (QEMUv8) or TEE core data-abort (HiKey960) when running `xtest 4002`, or "ftrace buffer too small" when running `optee_examples_hello_world` (HIKey960). It's possible that I did not clean my environment properly though.

FYI, I am building with `CFG_TA_FTRACE_SUPPORT=y CPPFLAGS_ta_arm64="-DFTRACE_BUF_SIZE=4096 -pg" CFG_ULIBS_MCOUNT=y`.

Reply:
I guess that the value of `ta_head` and `load_addr` happens to be the same, I'd rather use `load_addr` for clarity.

Reply:
ok

Reply:
Addressed comment.